text/0000-autoreborrow-traits.md

```diff
@@ -0,0 +1,468 @@
+    let result: Option<&mut U> = source.mutate();
+    result.observe();
+
+    // We want to enable the following calls to still work: this means that
+    // the first mutate() call must reborrow source instead of moving it.
+    // The following calls must also implicitly reborrow source as shared.
+    let result1: Option<&U> = source.observe();
+    let result2: Option<&U> = source.observe();
+    result1.observe();
+    result2.observe();
+    // The source must still be available for exlusive reborrowing afterwards.
+    source.mutate();
+}
+
+fn returning_derived_mut(source: Option<&mut T>) -> Option<&mut U> {
+    // This does work today, but mutate() must map() the option instead of eg.
+    // using as_deref_mut().
+    source.mutate()
+}
+
+fn returning_derived_mut(source: Option<&mut T>) -> Option<&U> {
+    // Note: This does not work today, and suggests using as_deref() which also
+    // doesn't work, citing "returns a value referencing data owned by the
+    // current function". Using map() and &* to explicitly reborrow the
+    // contained reference does work.
+    source.mutate()
+}
+```
+
+## Automatic reborrowing of custom reference types
+
+Custom reference types are as many as there are users, but a common refrain
+amongst them is that they usually hold one (though sometimes multiple) data
+fields and a PhantomData field for holding a lifetime in. Examples from above
+would now look like this:
+
+```rust
+/// Exclusive reborrowable reference to some data. Can be turned into a MyRef
+/// by reborrowing as shared.
+struct MyMut<'a>(...);
+/// Shared reborrowable reference to some data.
+struct MyRef<'a>(...);
+
+fn mixed_reborrowing(soure: MyMut) {
+    // Note: mutate() takes self by value, so MyMut would currently be moved
+    // here.
+    let result = source.mutate();
+    result.observe();
+
+    // We want to enable the following calls to still work: this means that
+    // the first mutate() call must reborrow source instead of moving it.
+    // The following calls must also implicitly reborrow source as shared.
+    let result1: Option<&U> = source.observe();
+    let result2: Option<&U> = source.observe();
+    result1.observe();
+    result2.observe();
+    // The source must still be available for exlusive reborrowing afterwards.
+    source.mutate();
+}
+
+fn returning_derived_mut(source: Option<&mut T>) -> Option<&mut U> {
+    // This does work today, but mutate() must map() the option instead of eg.
+    // using as_deref_mut().
+    source.mutate()
+}
+
+fn returning_derived_mut(source: Option<&mut T>) -> Option<&U> {
+    // Note: This does not work today, and suggests using as_deref() which also
+    // doesn't work, citing "returns a value referencing data owned by the
+    // current function". Using map() and &* to explicitly reborrow the
+    // contained reference does work.
+    source.mutate()
+}
+```
+
+A smattering of examples are:
```

Comment:
Bevy has a [handful](https://docs.rs/bevy/0.16.0-rc.5/bevy/app/trait.Plugin.html?search=reborrow) of examples. Probably the most interesting is the one on `Query` since semantically I don't think of that as a smart pointer.